src/config/identifier.rs

```diff
@@ -0,0 +1,168 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{borrow::Cow, fmt, str::FromStr};
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// A unique identifier for a configuration parameter.
+///
+/// Config identifiers must be:
+///
+/// * non-empty
+/// * ASCII printable
+/// * first character must be a letter
+/// * contain only letters, numbers, underscores, and hyphens
+///
+/// In general, config identifiers represent Rust package and Oxide service names.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
+#[serde(transparent)]
+pub struct ConfigIdent(Cow<'static, str>);
+
+impl ConfigIdent {
+    /// Creates a new config identifier at runtime.
+    pub fn new<S: Into<String>>(s: S) -> Result<Self, InvalidConfigIdent> {
+        let s = s.into();
+        Self::validate(&s)?;
+        Ok(Self(Cow::Owned(s)))
+    }
+
+    /// Creates a new config identifier from a static string.
+    pub fn new_static(s: &'static str) -> Result<Self, InvalidConfigIdent> {
+        Self::validate(s)?;
+        Ok(Self(Cow::Borrowed(s)))
+    }
+
+    /// Creates a new config identifier at compile time, panicking if the
+    /// identifier is invalid.
+    pub const fn new_const(s: &'static str) -> Self {
+        match Self::validate(s) {
+            Ok(_) => Self(Cow::Borrowed(s)),
+            Err(error) => panic!("{}", error.as_static_str()),
+        }
+    }
+
+    const fn validate(id: &str) -> Result<(), InvalidConfigIdent> {
+        if id.is_empty() {
+            return Err(InvalidConfigIdent::Empty);
+        }
+
+        let bytes = id.as_bytes();
+        if !bytes[0].is_ascii_alphabetic() {
+            return Err(InvalidConfigIdent::StartsWithNonLetter);
+        }
+
+        let mut bytes = match bytes {
+            [_, rest @ ..] => rest,
+            [] => panic!("already checked that it's non-empty"),
+        };
+        while let [next, rest @ ..] = &bytes {
+            if !(next.is_ascii_alphanumeric() || *next == b'_' || *next == b'-') {
+                break;
+            }
+            bytes = rest;
+        }
+
+        if !bytes.is_empty() {
+            return Err(InvalidConfigIdent::ContainsInvalidCharacters);
+        }
+
+        Ok(())
+    }
+
+    /// Returns the identifier as a string.
+    #[inline]
+    pub fn as_str(&self) -> &str {
+        &self.0
+    }
+}
+
+impl FromStr for ConfigIdent {
+    type Err = InvalidConfigIdent;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::new(s)
+    }
+}
+
+impl<'de> Deserialize<'de> for ConfigIdent {
```

Comment:
Why manually implement Deserialize but not Serialize? 

Is it worth adding a test that we can serialize -> deserialize -> re-serialize?

Reply:
`Deserialize` must go through validation, while serialization doesn't have to. Definitely worth writing a test for.

Reply:
Added tests and a comment explaining this.